types/lodash/common/lang.d.ts

```diff
@@ -1292,7 +1292,7 @@ declare module "../index" {
          * @param value The value to check.
          * @return Returns true if value is a plain object, else false.
          */
-        isPlainObject(value?: any): boolean;
+        isPlainObject(value?: any): value is Record<string | number | symbol, any>;
```

Comment:
Same, this produces incorrect typechecking for functions and classes

Reply:
@RyanCavanaugh Hi, I was trying to find some info on why isn't isPlainObject type-guarding for Record<>, can you give an example in which `isPlainObject(value?: any): value is Record<string | number | symbol, any>` would produce incorrect typechecking? `isPlainObject` is supposed to return `false` for functions and classes, shouldn't this proposed change work correctly?

Reply:
The reason this doesn't work, is that there are a number of types that are assignable to `Record<string | number | symbol, any>` that are _not_ plain objects per the lodash definition. That means that in the `true` branch of an `if` statement, we would allow types that `isPlainObject` wouldn't allow, and in the `false` branch we would remove types that `isPlainObject` would have rejected. You can see this by hovering over `value` in each of the branches in the following code on the playground:

https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXywGcAFaLVAeQCMArcDACgDcoJkQAueKVATwEouLNgiLwASuBwxgAHgDWIXjkTc+AGjW8AfAG4AUKEiwEYPIQzxh7LgEZ4AH3gYY7R-ADe8ABZYuAIgsYcgBzf3gAX3cGBn54AF5tKxwsYDinAFkoAAdZINDNfNQQpKcAbVtNACZNAGYAXQN9LFUGIlIocmo6TGZWdn44j314Uat+kF14AHpp+DNUFxwIJAgcAHd4QiheQnGRAj2Aci9fAKKwyOjYhKSmFLT3LNyLwpdQ0vgK6rr6o80qMhLFBMMhWEg0JhcPh1jhkBBgPA8BBeNwIGtNhhvAgAAanPzwQLvYrhCI4zRQQiEZAAW1CBAwe1QOGB+HIFh4YAQKm482gVIAdPoIvoQBBCAhhmN9uwprN5nglitEBitjs9tZRHscfYnC52OT4CAmCB8Fi4SFvPAce0yJRaPQcfBYfDEaxCDh4HAepYUOhsOZNM9CBTUG6YDB1cKgA

```ts
declare function isPlainObject(value: any): value is Record<keyof any, any>;
declare const value: 1 | true | { hi: "string" } | (() => void) | Map<string, string> | [1, 2, 3];

if (isPlainObject(value)) {
    value; // control flow says value is '{ hi: "string" } | (() => void) | Map<string, string> | [1, 2, 3]', 
           // but actual function would only allow the `{ hi: "string" }`, assuming its not an instance of a
           // class.
}
else {
    value; // control flow says value is `1 | true`, even though `isPlainObject` would also reject functions, 
           // Maps, and arrays
}
```

Reply:
I see, thank you so much! Seems like Typescript doesn't have the necessary tools to make a type guard here yet.